File: python/sat_test.py
#!/usr/bin/env python3

# Copyright (c) 2020 Nick Downing
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import imageio
import numpy
import sys
from gamma_decode import gamma_decode
from hsbk_to_rgb import hsbk_to_rgb

EXIT_SUCCESS = 0
EXIT_FAILURE = 1

RGB_RED = 0
RGB_GREEN = 1
RGB_BLUE = 2
N_RGB = 3

UVW_U = 0
UVW_V = 1
UVW_W = 2
N_UVW = 3

EPSILON = 1e-6

# inverse of the matrix calculated by ../prepare/UVW_to_rgb.py
rgb_to_UVW = numpy.array(
  [
    [9.0451048639000400e-02, 7.8430165104863919e-02, 3.9585452922802314e-02],
    [6.9958232931726871e-02, 2.3529049531459176e-01, 2.3751271753681389e-02],
    [4.0278982597054855e-02, 3.1372066041945568e-01, 1.6230035698348952e-01]
  ],
  numpy.double
)

if len(sys.argv) < 2:
  print(f'usage: {sys.argv[0]:s} image_out')
  print('image_out = name of PNG file to create (will be overwritten)')
  print('creates 361 x 376 x 3 image with 0..360 degrees by 1, 1500..9000 Kelvin by 20')
  sys.exit(EXIT_FAILURE)
image_out = sys.argv[1]

def hsbk_to_uv(hsbk):
  rgb = hsbk_to_rgb(hsbk)
  for i in range(N_RGB):
    rgb[i] = gamma_decode(rgb[i])
  UVW = rgb_to_UVW @ rgb
  return UVW[:UVW_W] / numpy.sum(UVW)

# find chromaticities of the hue space by 1 degree increments
hue_uv = numpy.stack(
  [
    hsbk_to_uv(numpy.array([1. * i, 1., 1., 6504.], numpy.double))
    for i in range(361)
  ],
  0
)

# find chromaticities of the Kelvin space by 20 degree increments
kelv_uv = numpy.stack(
  [
    hsbk_to_uv(numpy.array([0., 0., 1., 1500. + 20. * i], numpy.double))
    for i in range(376)
  ],
  0
)

# find chromaticities of the hue x Kelvin space @ saturation .5, then
# convert each chromaticity to a weighting between hue_uv and kelv_uv
image = numpy.zeros((376, 361), numpy.double)
for i in range(376):
  print(i, '/', 376)
  kelv = 1500. + 20. * i
  v0 = kelv_uv[i, :]
  for j in range(361):
    hue = 1. * j
    v1 = hue_uv[j, :] - v0
    uv = hsbk_to_uv(numpy.array([hue, .5, 1., kelv], numpy.double))
    w = ((uv - v0) @ v1) / (v1 @ v1)
    if w < 0.:
      w = 0.
    elif w > 1.:
      w = 1.
    image[i, j] = w

imageio.imwrite(image_out, numpy.round(image * 255.).astype(numpy.uint8))

File: c_float/sat_test.c
// Copyright (c) 2020 Nick Downing
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

#include <errno.h>
#include <math.h>
#include <png.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include "gamma_decode.h"
#include "hsbk_to_rgb.h"

#define RGB_RED 0
#define RGB_GREEN 1
#define RGB_BLUE 2
#define N_RGB 3

#define HSBK_HUE 0
#define HSBK_SAT 1
#define HSBK_BR 2
#define HSBK_KELV 3
#define N_HSBK 4

#define UV_u 0
#define UV_v 1
#define N_UV 2

#define UVW_U 0
#define UVW_V 1
#define UVW_W 2
#define N_UVW 3

#define EPSILON 1e-6f

// inverse of the matrix calculated by ../prepare/UVW_to_rgb.py
float rgb_to_UVW[N_UVW][N_RGB] = {
  {9.04510486e-02f, 7.84301651e-02f, 3.95854529e-02f},
  {6.99582329e-02f, 2.35290495e-01f, 2.37512718e-02f},
  {4.02789826e-02f, 3.13720660e-01f, 1.62300357e-01f}
};

void hsbk_to_uv(const float *hsbk, float *uv) {
  float rgb[N_RGB];
  hsbk_to_rgb(hsbk, rgb);

  for (int i = 0; i < N_RGB; ++i)
    rgb[i] = gamma_decode(rgb[i]);

  float UVW[N_UVW];
  for (int i = 0; i < N_UVW; ++i) {
    float v = 0.f;
    for (int j = 0; j < N_RGB; ++j)
      v += rgb_to_UVW[i][j] * rgb[j];
    UVW[i] = v;
  }

  float L = 0.f;
  for (int i = 0; i < N_UVW; ++i)
    L += UVW[i];
  uv[UV_u] = UVW[UVW_U] / L;
  uv[UV_v] = UVW[UVW_V] / L;
}

int main(int argc, char **argv) {
  if (argc < 2) {
    printf(
      "usage: %s image_out\n"
        "image_out = name of PNG file to create (will be overwritten)\n"
        "creates 361 x 376 x 3 image with 0..360 degrees by 1, 1500..9000 Kelvin by 20\n",
      argv[0]
    );
    exit(EXIT_FAILURE);
  }
  char *image_out = argv[1];

  // find chromaticities of the hue space by 1 degree increments
  float hue_uv[361][N_UV];
  for (int i = 0; i < 361; ++i) {
    float hsbk[N_HSBK] = {1.f * i, 1.f, 1.f, 6504.f};
    hsbk_to_uv(hsbk, hue_uv[i]);
  }

  // find chromaticities of the Kelvin space by 20 degree increments
  float kelv_uv[376][N_UV];
  for (int i = 0; i < 376; ++i) {
    float hsbk[N_HSBK] = {0.f, 0.f, 1.f, 1500.f + 20.f * i};
    hsbk_to_uv(hsbk, kelv_uv[i]);
  }

  // find chromaticities of the hue x Kelvin space @ saturation .5, then
  // convert each chromaticity to a weighting between hue_uv and kelv_uv
  float image[376][361];
  for (int i = 0; i < 376; ++i) {
    //printf("%d / 376\n", i);
    float kelv = 1500.f + 20.f * i;
    float v0[N_UV] = {
      kelv_uv[i][UV_u],
      kelv_uv[i][UV_v]
    };
    for (int j = 0; j < 361; ++j) {
      float hue = 1.f * j;
      float v1[N_UV] = {
        hue_uv[j][UV_u] - v0[UV_u],
        hue_uv[j][UV_v] - v0[UV_v]
      };
      float hsbk[N_HSBK] = {hue, .5f, 1.f, kelv};
      float uv[N_UV];
      hsbk_to_uv(hsbk, uv);
      float w = (
        (uv[UV_u] - v0[UV_u]) * v1[UV_u] +
          (uv[UV_v] - v0[UV_v]) * v1[UV_v]
      ) / (
        v1[UV_u] * v1[UV_u] +
          v1[UV_v] * v1[UV_v]
      );
      if (w < 0.f)
        w = 0.f;
      else if (w > 1.f)
        w = 1.f;
      image[i][j] = w;
    }
  }

  FILE *fp = fopen(image_out, "wb");
  if (fp == NULL) {
    perror(image_out);
    exit(EXIT_FAILURE);
  }

  png_structp png_ptr = png_create_write_struct(
    PNG_LIBPNG_VER_STRING,
    NULL,
    NULL,
    NULL
  );
  if (png_ptr == NULL) {
    perror("png_create_write_struct()");
    exit(EXIT_FAILURE);
  }

  png_infop info_ptr = png_create_info_struct(png_ptr);
  if (info_ptr == NULL) {
    perror("png_create_info_struct()");
    exit(EXIT_FAILURE);
  }

  if (setjmp(png_jmpbuf(png_ptr))) {
    perror("setjmp()");
    exit(EXIT_FAILURE);
  }
  png_init_io(png_ptr, fp);

  png_set_IHDR(
    png_ptr,
    info_ptr,
    361,
    376,
    8,
    PNG_COLOR_TYPE_GRAY,
    PNG_INTERLACE_NONE,
    PNG_COMPRESSION_TYPE_BASE,
    PNG_FILTER_TYPE_BASE
  );

  png_write_info(png_ptr, info_ptr);

  png_byte pixels[376][361];
  for (int i = 0; i < 376; ++i)
    for (int j = 0; j < 361; ++j)
      pixels[i][j] = (png_byte)roundf(image[i][j] * 255.f);

  png_byte *rows[376];
  for (int i = 0; i < 376; ++i)
    rows[i] = pixels[i];

  png_write_image(png_ptr, rows);

  png_write_end(png_ptr, NULL);
  fclose(fp);

  return EXIT_SUCCESS;
}
